core-api/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,263 @@
+use std::{fmt::Debug, sync::Arc};
+
+/// Implementors of this trait are expected to be defined in each language's bridge.
+/// The implementor is responsible for the allocation/instantiation of new metric meters which
+/// Core has requested.
+pub trait CoreMeter: Send + Sync + Debug {
+    fn new_attributes(&self, attribs: MetricsAttributesOptions) -> MetricAttributes;
+    fn counter(&self, name: &str) -> Arc<dyn Counter>;
```

Comment:
I can't. I actually had quite the struggle with the type system on trying to get this to be as little overhead as possible. Traits cannot return existential types (can't return `impl Counter`) yet. I would need to make the trait generic, or have an associated type, but that would infect the entire rest of the code with generics just for this purpose which would be no bueno.

Thus, these methods have to return some kind of pointer, so might as well be `Arc` here since `Box` is just going to get shoved in one anyway

Reply:
The only other option would make the instrument types also be enums, which is another spot where there can be a mis-match between the attribute type and the instrument type, which is sorta sad. But, given that can already happen, I'd probably be fine with enum-ing them too.

Reply:
I don't mean enum instrument, I mean enum counter which is either otel-in-core backed, or lang-backed. 

The lang-backed version would essentially do what you suggest there, except it would store an internal `u64` or something which would be used for a pre-hashed or indexed lookup lang side rather than hashing the name every time.